### ISSUE TITLE
Make calibre use the system theme

### ISSUE THREAD

**Issue by xcffl** [flathub/com.calibre_ebook.calibre#34]
As explained in #2, calibre doesn't really use the system theme but use the built-in one which is not good-looking.

There are definitely differences. Below are examples. The upper ones respect system's Adwaita theme whereas the lower ones use the built-in theme.

## Main window
![calibre_use_system_theme_1](https://user-images.githubusercontent.com/2216902/75736345-e97ba380-5d37-11ea-98ec-67a024aaa792.png)
![Calibre without system theme  - main window](https://user-images.githubusercontent.com/2216902/75736598-a110b580-5d38-11ea-8393-3628475402f2.png)

## Edit metadata
![calibre_use_system_theme_2](https://user-images.githubusercontent.com/2216902/75736352-f00a1b00-5d37-11ea-8c52-18f6d5dc8ae8.png)
![Calibre without system theme  - edit metadata](https://user-images.githubusercontent.com/2216902/75736226-96095580-5d37-11ea-98d5-ab7cbce660e5.png)

It's trivial to add `CALIBRE_USE_SYSTEM_THEME=true` env variable but makes Calibre look much better.

**Comment by TingPing** [comment]
Generally flatpaks respect upstream choices and considering Calibre hides this behind an env var defaulting off I think they want this behavior.

**Comment by xcffl** [comment]
> Generally flatpaks respect upstream choices and considering Calibre hides this behind an env var defaulting off I think they want this behavior.

Actually it's not what they want but because of bugs it introduce. With Flatpak I think it's not a problem.

**Comment by Erick555** [comment]
What bugs and why they aren't problem in flatpak?

**Comment by xcffl** [comment]
> What bugs and why they aren't problem in flatpak? If neither upstream or any distro default to system theme then why flatpak should give different experience for users

Flatpak from my understanding fixes the runtime environment which address [the problem](https://www.mobileread.com/forums/showthread.php?t=254507#4) of "Qt 5 GTK plugin was causing crashes on many systems I tested with, so it is left out of the calibre bundle," as @kovidgoyal said (it would be better if he may confirm this).  So we just need a _working plugin_.

But it could simpler. Actually if [adwaita-qt](https://github.com/MartinBriza/adwaita-qt) is packaged into a Flatpak theme, it doesn't need any style plugin, since [it provides Qt style directly](https://wiki.archlinux.org/index.php/Uniform_look_for_Qt_and_GTK_applications#Styles_for_both_Qt_and_GTK). So no problem for that (and it's tested). 

If you think it's reasonable, we may think of make it happen.

**Comment by kovidgoyal** [comment]
gtk themes constantly break backwards compatibility and cause hard to debug crashes/misbehavior. I dont have the time/patience to track those down. As such calibre is not going to enable the system theme by default.

**Comment by xcffl** [comment]
> gtk themes constantly break backwards compatibility and cause hard to debug crashes/misbehavior. I dont have the time/patience to track those down. As such calibre is not going to enable the system theme by default.

I understand that and appretiate your work. But it doesn't seem to be any backwards compatibility issue in flatpak world? AFAIK flatpak is built to address this problem.

**Comment by xcffl** [comment]
Also, it won't and shouldn't give your additional workload as most things happens here.

**Comment by kovidgoyal** [comment]
Well if you are willing to guarantee compatibility and debug any
resulting crashes yourself, feel free to enable it in the flatpak
distro. I am not going to complain.

On Tue, Mar 03, 2020 at 07:21:13AM -0800, xcffl wrote:
> > gtk themes constantly break backwards compatibility and cause hard to debug crashes/misbehavior. I dont have the time/patience to track those down. As such calibre is not going to enable the system theme by default.
> 
> I understand that and appretiate your work. But it doesn't seem to be any backwards compatibility issue in flatpak world? AFAIK flatpak is built to address this problem.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/flathub/com.calibre_ebook.calibre/issues/34#issuecomment-594005744

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Comment by xcffl** [comment]
> Well if you are willing to guarantee compatibility and debug any resulting crashes yourself, feel free to enable it in the flatpak distro. I am not going to complain.
> […](#)
> On Tue, Mar 03, 2020 at 07:21:13AM -0800, xcffl wrote: > gtk themes constantly break backwards compatibility and cause hard to debug crashes/misbehavior. I dont have the time/patience to track those down. As such calibre is not going to enable the system theme by default. I understand that and appretiate your work. But it doesn't seem to be any backwards compatibility issue in flatpak world? AFAIK flatpak is built to address this problem. -- You are receiving this because you were mentioned. Reply to this email directly or view it on GitHub: [#34 (comment)](https://github.com/flathub/com.calibre_ebook.calibre/issues/34#issuecomment-594005744)
> --
> _____________________________________ Dr. Kovid Goyal https://www.kovidgoyal.net https://calibre-ebook.com
> _____________________________________

That's great!

**Comment by Erick555** [comment]
Well, currently calibre flatpak uses official upstream binaries and it's not possible to fix issues like [this one](https://github.com/flathub/com.calibre_ebook.calibre/issues/30) without upstream help. For this reason I think losing upstream support only because someone found it too hard to set one env variable themselves which they have to set in every other platform anyway isn't a wise choice.

**Comment by xcffl** [comment]
I think he only talks about the bug **introduced** by that option. This can be determined by just turning off that option.

BTW the adwaita kstyle is already there as `org.kde.KStyle.Adwaita`, so it seems not so hard to implement this feature.

**Comment by xcffl** [comment]
Well, so simply using `org.kde.Platform` as runtime and `--env=CALIBRE_USE_SYSTEM_THEME=true`make it. It's easy to debug and at least the upstream will not be unhappy about switching to `org.kde.Platform`.

**Comment by Erick555** [comment]
kde platform is ~228MB bigger than freedesktop platform and all additional content is useless for calibre which uses their own qt so this is another major obstacle.

Also what if user uses different theme than adwaita?

**Comment by ghisvail** [comment]
> kde platform is ~228MB bigger than freedesktop platform and all additional content is useless for calibre which uses their own qt so this is another major obstacle.

Imo, the improved ethetics do not outweigh the cost of the increased size of the runtime.

Rather, I believe efforts should be spent on a proper source build (#17), which may provide better opportunities for integration with other themes.

**Comment by xcffl** [comment]
> kde platform is ~228MB bigger than freedesktop platform and all additional content is useless for calibre which uses their own qt so this is another major obstacle.

Well, I think in most cases people install all three SDKs. However if they don't, I agree with you they may need that 200+MB (still thinking of a better solution). But I notice that a lot of Flatpaks are built against freedesktop SDK, which means probably they will not get native experience (meaning they will look wired whether you use KDE or Gnome)

> Also what if user uses different theme than adwaita?
After some investigation, it seems that, under GNOME:
- if their theme has a Flatpak, they will fallback to GTK2 version of that theme
- if it doesn't, they will fallback to Adwaita
Under KDE:
- it will be fallback to Breeze, or the corresponding KStyle theme as same as under GNOME.

**Comment by xcffl** [comment]
> Rather, I believe efforts should be spent on a proper source build (#17), which may provide better opportunities for integration with other themes.

The discussion there even inspires that we may bundle  `org.kde.KStyle.Adwaita` and `org.gtk.Gtk3theme.Breeze` directly into Calibre Flatpak, at the cost of 394.2 kB + 18.0 MB.

**Comment by Erick555** [comment]
> But I notice that a lot of Flatpaks are built against freedesktop SDK, which means probably they will not get native experience (meaning they will look wired whether you use KDE or Gnome)

On KDE all my installed flatpaks that use freedesktop SDK runtime (some of them use gtk toolkit and some qt one) are native looking and indistinguishable from non-flatpak apps. This includes calibre with `CALIBRE_USE_SYSTEM_THEME=true` set. I don't know how it looks in other DE though.

**Comment by xcffl** [comment]
> On KDE all my installed flatpaks that use freedesktop SDK runtime (some of them use gtk toolkit and some qt one) are native looking and indistinguishable from non-flatpak apps. This includes calibre with `CALIBRE_USE_SYSTEM_THEME=true` set. I don't know how it looks in other DE though.

Perhaps you've installed kde runtime?

**Comment by xcffl** [comment]
> > But I notice that a lot of Flatpaks are built against freedesktop SDK, which means probably they will not get native experience (meaning they will look wired whether you use KDE or Gnome)
> 
> On KDE all my installed flatpaks that use freedesktop SDK runtime (some of them use gtk toolkit and some qt one) are native looking and indistinguishable from non-flatpak apps. This includes calibre with `CALIBRE_USE_SYSTEM_THEME=true` set. I don't know how it looks in other DE though.

Suprising I can't really find an app built upon Qt but only uses freedesktop SDK... But that makes sense, unless they are prebuilt or add Qt as their module during the building process.

Or may you list any app that isn't in the above cases if there is any?

**Comment by Erick555** [comment]
> Perhaps you've installed both the kde & gnome runtime?

I have kde runtime (not gnome) installed but thy it would matter for apps which don't use it?

> Suprising I can't really find an app built upon Qt but only uses freedesktop SDK... But that makes sense, unless they are prebuilt or add Qt as their module during the building process.

I have installed two. One is calibre, second one is https://github.com/flathub/com.elsevier.MendeleyDesktop . Both are prebuilt apps for whom additional qt in runtime may cause incompatibility conflicts and brings no benefits.

**Comment by xcffl** [comment]
>  I have kde runtime (not gnome) installed but thy it would matter for apps which don't use it?

That is for those who use kde/gnome runtime. They actually use the corresponding theme packages.

> I have installed two. One is calibre, second one is https://github.com/flathub/com.elsevier.MendeleyDesktop . Both are prebuilt apps for whom additional qt in runtime may cause incompatibility conflicts and brings no benefits.

As expected Mendeley looks so legacy on GNOME. I will test what it looks like on KDE.

So if #17 passes, there will not be any trade-off regarding additional size or incompatibility, since we will have qt binary during the building process (and that's all we need). There is no freedesktop sdk based GTK apps that don't look native on KDE. 

Maybe we can do the native theming case-by-case. If an app is stable to have native theming, we make it looks pleasant. Vice versa.
Additionally, to clarify, I don't believe outside environment will have big impact to make apps running in the **same containerized environment** show divergent behaviors. Therefore, if it's well-tested when packaging and have no error, we should expect it will be OK on end-user's computer.

**Comment by TingPing** [comment]
> The discussion there even inspires that we may bundle org.kde.KStyle.Adwaita (for GNOME users) and org.gtk.Gtk3theme.Breeze (for KDE users) directly into Calibre Flatpak, at the cost of 394.2 kB + 18.0 MB.

One is a Qt theme, one is a GTK theme. You don't need both.

**Comment by xcffl** [comment]
> > The discussion there even inspires that we may bundle org.kde.KStyle.Adwaita (for GNOME users) and org.gtk.Gtk3theme.Breeze (for KDE users) directly into Calibre Flatpak, at the cost of 394.2 kB + 18.0 MB.
> 
> One is a Qt theme, one is a GTK theme. You don't need both. (Qt doesn't load GTK3 themes in any form to my knowledge)

That's true. So it'll be extra 18MB for Qt apps and 393.2 kB for GTK+ apps.

**Comment by LinAGKar** [comment]
> kde platform is ~228MB bigger than freedesktop platform and all additional content is useless for calibre which uses their own qt so this is another major obstacle.

It's shouldn't be bundling its own Qt in the first place, that's the whole point of the runtime system.

> The discussion there even inspires that we may bundle `org.kde.KStyle.Adwaita` (for GNOME users) and `org.gtk.Gtk3theme.Breeze` (for KDE users) directly into Calibre Flatpak, at the cost of 394.2 kB + 18.0 MB.

That seems like a hack that will break if you don't one of those specific themes. Also, would that use the color scheme set in the KDE settings? Because Qt apps using the KDE runtime do (I've tested with org.qgis.qgis).

IMO, it should use the KDE runtime, which would probably be installed anyway if flatpaks become heavily used, in which case it will work with system colors and kstyle themes.

**Comment by Erick555** [comment]
> It's shouldn't be bundling its own Qt in the first place, that's the whole point of the runtime system.

The fact is it does and will do in foreseeable future, see https://github.com/flathub/com.calibre_ebook.calibre/issues/17

> That seems like a hack that will break if you don't one of those specific themes. Also, would that use the color scheme set in the KDE settings? Because Qt apps using the KDE runtime do (I've tested with org.qgis.qgis).

If you use theme that isn't available in flatpak then it will never work there. Custom color scheme works right now in calibre flatpak.

> IMO, it should use the KDE runtime, which would probably be installed anyway if flatpaks become heavily used, in which case it will work with system colors and kstyle themes.

For the reasons stated above it's pointless right now and may be harmful if there is a mismatch between runtime qt and calibre own qt versions.

**Comment by LinAGKar** [comment]
> The fact is it does and will do in foreseeable future, see #17

It relying on Qt internals is its own problem, but OK, that's not gonna be solved here.

> If you use theme that isn't available in flatpak then it will never work there.

Though it will if the theme is packaged for flatpak, which is not the case the app only uses themes bundled with it.

> Custom color scheme works right now in calibre flatpak.

Not in my experience. I can change the color scheme in the KDE settings, and that changes qgis, but not calibre.

**Comment by Erick555** [comment]
> Though it will if the theme is packaged for flatpak, which is not the case the app only uses themes bundled with it.

Yes, it must be packaged for flatpak so you can't use any theme but I think there is only one QT theme available right now (org.kde.KStyle.Adwaita)

> Not in my experience. I can change the color scheme in the KDE settings, and that changes qgis, but not calibre.

Did you set `CALIBRE_USE_SYSTEM_THEME=true` variable?

**Comment by LinAGKar** [comment]
> Did you set CALIBRE_USE_SYSTEM_THEME=true variable?

Yes, I tried with `flatpak override --env=CALIBRE_USE_SYSTEM_THEME=true com.calibre_ebook.calibre` and `CALIBRE_USE_SYSTEM_THEME=true flatpak run com.calibre_ebook.calibre`, but it still does not respect the system color scheme.

**Comment by Erick555** [comment]
Is this on kde?

**Comment by LinAGKar** [comment]
>Is this on kde?

Yes, it is, on openSUSE Tumbleweed. I set it to the Breeze Dark color scheme in the KDE color settings, and that works for calibre from RPM, and the qgis flatpak, but not for the calibre flatpak.

**Comment by Erick555** [comment]
I don't know then, for me it just works in Arch Linux, kde:
![calibre](https://user-images.githubusercontent.com/37542552/84141898-68712e00-aa43-11ea-9b8f-ca8d495bf5c3.jpg)

**Comment by alxlg** [comment]
> > Did you set CALIBRE_USE_SYSTEM_THEME=true variable?
> 
> Yes, I tried with `flatpak override --env=CALIBRE_USE_SYSTEM_THEME=true com.calibre_ebook.calibre` and `CALIBRE_USE_SYSTEM_THEME=true flatpak run com.calibre_ebook.calibre`, but it still does not respect the system color scheme.

Same here. KDE Neon, org.kde.Platform installed, QtWidget theme set to Breeze. It works with VLC, but not with Calibre.

@Erick555 Do you have any other Flatpak package installed?

**Comment by Erick555** [comment]
Yes, for me it works everywhere.

**Comment by Erick555** [comment]
@LinAGKar @alxlg sorry for this taking so long but I finally found what is missing: you may try running `flatpak run --filesystem=xdg-config/kdeglobals:ro com.calibre_ebook.calibre` and this should make color schemes work (at least on kde plasma session and assuming it works for other flatpaks for you)

**Comment by LinAGKar** [comment]
That does work, though only that's not a permanent solution.

**Comment by Erick555** [comment]
@LinAGKar this will be permanent solution when https://github.com/flathub/com.calibre_ebook.calibre/pull/59 is merged

**Comment by ghisvail** [comment]
@Erick555 @LinAGKar is there anything left actionable here? Otherwise, let's close it.

**Comment by Erick555** [comment]
There was a request for passing `CALIBRE_USE_SYSTEM_THEME=true` by default but I agree with a [comment](https://github.com/flathub/com.calibre_ebook.calibre/issues/34#issuecomment-593756669) that we should honor upstream choice to leave it as user opt-in.

**Comment by ghisvail** [comment]
> we should honor upstream choice to leave it as user opt-in.

Agreed. Leaving it as `wontfix` then.

**Comment by xcffl** [comment]
So, for anyone who want to do this by their own, you may run `flatpak run --env=CALIBRE_USE_SYSTEM_THEME=true --runtime=org.kde.Platform --runtime-version=5.15 com.calibre_ebook.calibre`. And you may even save the override by `flatpak override --env=CALIBRE_USE_SYSTEM_THEME=true`. Unfortunately there is no way to override the runtime permanently

**Comment by ghisvail** [comment]
Thanks for your input 👍

**Comment by rokejulianlockhart** [comment]
@xcffl, the issue that https://github.com/flathub/org.kde.WaylandDecoration.QGnomePlatform-decoration/issues/4 depends upon was purportedly remediated by https://github.com/DavidoTek/ProtonUp-Qt/issues/98#event-7376680589, so does that not remediate this?

**Comment by Erick555** [comment]
which kde runtime you test, 6.3?

**Comment by Erick555** [comment]
The amount of patches will change over time. The qt version calibre uses will change over time. Calibre supports only Qt libraries provided on their own.

**Comment by kovidgoyal** [comment]
You want me to add zstd to the sources.json as a calibre dependency? There's already pyzstd so including zstd as a separate library should be fine, one will need to make pyzstd build against it.

**Comment by Erick555** [comment]
I think bringing ztsd to calibre is orthogonal to the issue - it won't change anything without additional user hacks.

The real fix would be to add `adwaita-qt` and `QGnomePlatform` to calibre. That should fix app look in gnome.

**Comment by kovidgoyal** [comment]
I am definitely not going to be shipping GNOME or Adwaita with calibre
binaries.

**Comment by Erick555** [comment]
Technically libadwaitaqt6 doesn't depend on anything GNOME.

QGnomePlatform [uses private Qt abi](https://github.com/FedoraQt/QGnomePlatform#how-to-compile) so its compatibility between runtime Qt and alibre Qt can be only accidental.

**Comment by Erick555** [comment]
The Qt compatibility problem isn't specific to flapak, same thing occurs with distro Qt.

> Not quite true, unless "accidental" means highly likely

Upstream states **This library will have to be recompiled with every Qt update.** It's only pure coincidence that Calibre will use exact same Qt version as flatpak or bare-meal distro and with exact same features enabled. Today it's zstd tomorrow it will be something else. It's cat and mouse game. Only calibre sporting theming itself can guarantee things will work in the future.

> That is, provide support for approaching the behavior as if it were built from source, at least at a minimum level, as long as there's a real user need?

For 99% users of calibre and flatpak adding zstd doesn't change theming situation at all. I wouldn't call this _support_.

**Comment by Erick555** [comment]
You can see how complicated source build is [here](https://github.com/flathub/com.calibre_ebook.calibre/pull/53). This needs manual intervention for every update. Right now you can have flatpak calibre update the day after it's released. With source build it would be months (assuming someone would want to maintain it at all).

**Comment by ghisvail** [comment]
> You can see how complicated source build is [here](https://github.com/flathub/com.calibre_ebook.calibre/pull/53). This needs manual intervention for every update. Right now you can have flatpak calibre update the day after it's released. With source build it would be months (assuming someone would want to maintain it at all).

Have a look at the "Updating this Flatpak" section specifically. Unless some serious efforts are spent to **fully** automate translation of upstream's manifest to Flatpak, manual intervention will result in slower update cycles and potential human mistakes in the process. As a maintainer of Electron-based Flatpak packages, I have been facing this issue many times.

So far we have favored practicality and speed (and our sanity :sweat_smile:)  by repacking upstream binaries, at the expense of a few integration gotchas.

**Comment by pjkaufman** [comment]
Good morning. I am trying to get Calibre to respect the QT style that I have installed, but for some reason I am not really able to get it to do so. I am probably doing something wrong, so feel free to let me know if I am getting of track or if there is a simple way to do this with system themes that are not in flatpak.

To preface, I am on wayland via Sway 1.7 and Pop OS! 22.04.

To start I set `export CALIBRE_USE_SYSTEM_THEME=true` in my profile to have it trickle down into the calibre flatpak (if that is not enough I also added the override to calibre via `flatpak override --env=CALIBRE_USE_SYSTEM_THEME=true`). This does not seem to have any effect on the styling in the calibre flatpak.

So I did the following:
1. Installed the kvantum flatpak (`org.kde.KStyle.Kvantu org.kde.KStyle.Kvantum`)
2. Installed `qt5-style-kvantum` and `qt5-style-kvantum-themes`
3. Ran `kvantummanager` and installed my theme (in this case it was the [catpuccin theme for qt5ct](https://github.com/catppuccin/qt5ct) and set it to the default theme.
4. Set the following export in my bash profile `export QT_QPA_PLATFORMTHEME=kvantum`
5. Set the filesystem access for all flatpaks to see the Kvantum config: `flatpak override --filesystem="$HOME/.config/Kvantum"`

However I am not seeing calibre respect the catppuccin theme or an option to set the theme to catppuccin under the apprearance settings.

I have also tried setting `export QT_STYLE_OVERRIDE=kvantum` in my bash profile and I get a message saying:
``` sh
QApplication: invalid style override 'kvantum' passed, ignoring it.
	Available styles: Windows, Fusion
```
I am not sure why it would ignore kvantum since it is an installed flatpak and it has access to the kvantum configs.

Is there something I am doing wrong here? I am newer to how to style QT apps, so it would be great if I could get some pointers on steps to debug things like this. Thanks for all the help and work that has been put into this!

**Comment by Erick555** [comment]
In general this issue is about theming being unsupported. There are [hacks](https://github.com/flathub/com.calibre_ebook.calibre/issues/34#issuecomment-776037458) which may or may not work for you.

**Comment by pjkaufman** [comment]
> In general this issue is about theming being unsupported. There are [hacks](https://github.com/flathub/com.calibre_ebook.calibre/issues/34#issuecomment-776037458) which may or may not work for you.

I think I am a bit lost. From what I read in this thread, setting `CALIBRE_USE_SYSTEM_THEME=true` supposedly does allow the use of system theming from the docs and the discussion here. However it does not seem to do so in my case, which seems to be a trend here.

Is there a specific version of qt that I need to account for with the theming (i.e. qt6 or qt5)? I ask because my understanding is that `qt6` is not yet available on most package managers which would make styling for it pretty hard since that would require manually building the tools in question. Any tips on debugging this myself or a better place to ask this if this is not the place to do so?

**Comment by Erick555** [comment]
Using host color schemes will work but qt themes not.

In flatpak qt themes are tied to `org.kde.Platform//xx` runtime but since calibre bundles its own qt it uses `org.freedesktop.Platform//xx` runtime instead. The bundled qt version is changed with its own pace so using matching runtime version is cat & mouse game.

I think calibre right now uses qt 6.4 so you would need to install `org.kde.Platform//6.4` (it's EOL on flathub) then install relevant theme matching 6.4 branch then running:
`flatpak run --env=CALIBRE_USE_SYSTEM_THEME=true --runtime=org.kde.Platform --runtime-version=6.4 com.calibre_ebook.calibre` each time.

**Comment by pjkaufman** [comment]
Thanks for the fast response. I see that you are differentiating between host color schemes and qt themes. Is there a place to read up on that difference? I believe that qt themes are generally used by KDE while GTK themes are used by Gnome. I would assume that on a Gnome Sway setup that means GTK is the host color scheme, but I am not seeing that actually happen in calibre with use system theme set..

**Comment by Erick555** [comment]
Calibre is qt app so gtk settings shouldn't matter for it. I'm on kde with default theme breeze and when I set any color scheme in system settings calibre will follow it without additional setup. Buttons and decorations are slightly different but for me it's hardly noticeable. 

I guess for other desktop environments even the color schemes may not work without additional setup.

**Comment by pjkaufman** [comment]
Thanks for the help with understanding this! I will have to see about testing some things around this.
